### PR TITLE
Add support for writer version 7 in Delta Lake

### LINKED ISSUE #15873: Delta Lake: Writer version 7 support
Version 7 introduced table features: https://github.com/delta-io/delta/blob/master/PROTOCOL.md#table-features

Note: We should check writer features in `vacuum` procedure.

### PR DESCRIPTION
**Pull request by ebyhr**
## Description

Fixes #15873 

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Delta Lake
* Add support for writer version 7. ({issue}`15873`)
```

### CODE REVIEW

**Comment by findinpath** [comment]
Build is :red_circle: 

```
Error:  Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.11.0:compile (default-compile) on project trino-product-tests: Compilation failure
Error:  /home/runner/work/trino/trino/testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:[131,80] cannot find symbol
Error:    symbol:   variable ENGLISH
Error:    location: class io.trino.tests.product.deltalake.TestDeltaLakeColumnMappingMode
```

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java [hunk 1]
```diff
@@ -200,6 +205,46 @@ public void testCheckConstraintMergeCompatibility()
...
+    public void testCheckConstraintWriterFeature()
+    {
+        String tableName = "test_check_constraint_writer_feature_" + randomNameSuffix();
+
+        onDelta().executeQuery("CREATE TABLE default." + tableName +
+                "(a INT) " +
+                "USING DELTA " +
+                "LOCATION 's3://" + bucketName + "/databricks-compatibility-test-" + tableName + "'" +
+                "TBLPROPERTIES ('delta.minWriterVersion'='7')");
+        onDelta().executeQuery("ALTER TABLE default." + tableName + " ADD CONSTRAINT a_constraint CHECK (a > 1)");
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java:219]: Could you pls add as well a table a table level check constraint next to column level check constraints to ensure that they are also supported

https://github.com/trinodb/trino/blob/945cb416fda0c8e2ffc2ea94a96f11c833c1cbbf/testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java#L240

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeWriteDatabricksCompatibility.java [hunk 2]
```diff
@@ -395,17 +396,42 @@ private void testVacuumRemoveChangeDataFeedFiles(Consumer<String> vacuumExecutor
...
+        String tableName = "test_vacuum_unsupported_writer_feature_" + randomNameSuffix();
+        String directoryName = "databricks-compatibility-test-" + tableName;
+
+        onDelta().executeQuery("CREATE TABLE default." + tableName +
+                "(a INT)" +
+                "USING DELTA " +
+                "LOCATION '" + ("s3://" + bucketName + "/" + directoryName) + "'" +
+                "TBLPROPERTIES ('delta.minWriterVersion' = '7', 'delta.enableDeletionVectors' = true)");
         try {
+            // TODO Improve error message
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeWriteDatabricksCompatibility.java:432]: :+1:

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointWriter.java [hunk 3]
```diff
@@ -113,7 +113,7 @@ public void write(CheckpointEntries entries, TrinoOutputFile outputFile)
         RowType metadataEntryType = checkpointSchemaManager.getMetadataEntryType();
         RowType protocolEntryType = checkpointSchemaManager.getProtocolEntryType(protocolEntry.getReaderFeatures().isPresent(), protocolEntry.getWriterFeatures().isPresent());
         RowType txnEntryType = checkpointSchemaManager.getTxnEntryType();
-        RowType addEntryType = checkpointSchemaManager.getAddEntryType(entries.getMetadataEntry(), writeStatsAsJson, writeStatsAsStruct);
+        RowType addEntryType = checkpointSchemaManager.getAddEntryType(entries.getMetadataEntry(), entries.getProtocolEntry(), writeStatsAsJson, writeStatsAsStruct);
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/checkpoint/CheckpointWriter.java:116]: Given that your PR is equipped only with a test in `TestDeltaLakeAlterTableCompatibility` which exercises checkpoint writing for the situation when a table is written with writer version `7`, please add an extra test to ensure this code is properly covered in `TestDeltaLakeDatabricksCheckpointsCompatibility`.

**Comment by findinpath** [comment]
LGTM % comments

**Comment by ebyhr** [comment]
Just rebased on master to resolve conflicts.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 4]
```diff
@@ -694,11 +697,13 @@ public Iterator<TableColumnsMetadata> streamTableColumns(ConnectorSession sessio
                             return Stream.of();
                         }
                         String tableLocation = metastoreTable.get().location();
-                        MetadataEntry metadata = transactionLogAccess.getMetadataEntry(getSnapshot(table, tableLocation, session), session);
+                        TableSnapshot snapshot = getSnapshot(table, tableLocation, session);
+                        MetadataEntry metadata = transactionLogAccess.getMetadataEntry(snapshot, session);
+                        ProtocolEntry protocol = transactionLogAccess.getProtocolEntry(session, snapshot);
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:702]: Can this end up doing second time same IO as done just above (in `transactionLogAccess.getMetadataEntry`)?
Looking at the code this may happen if snapshot has a checkpoint attached.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 5]
```diff
@@ -2160,16 +2178,24 @@ private boolean allowWrite(ConnectorSession session, DeltaLakeTableHandle tableH
 
     private void checkWriteSupported(ConnectorSession session, DeltaLakeTableHandle handle)
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2179]: session is now unused

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 6]
```diff
@@ -549,6 +597,11 @@ public static Set<String> unsupportedReaderFeatures(Set<String> features)
         return Sets.difference(features, SUPPORTED_READER_FEATURES);
     }
 
+    public static Set<String> unsupportedWriterFeatures(Set<String> features)
+    {
+        return Sets.difference(features, SUPPORTED_WRITER_FEATURES);
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:602]: A static method works now, but what if we support for a new writer feature is conditional based on config or session?

I think it would be safer to have this as a instance method in DeltaLakeMetadata.
I see however this is same as in `unsupportedReaderFeatures`, so it's already committed state. Fine to keep for now.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 7]
```diff
@@ -491,17 +530,26 @@ private static String getGeneratedColumnExpressions(JsonNode node)
...
+        if (protocolEntry.supportsWriterFeature() && !protocolEntry.writerFeaturesContains("checkConstraints")) {
+            return ImmutableMap.of();
+        }
         return metadataEntry.getConfiguration().entrySet().stream()
                 .filter(entry -> entry.getKey().startsWith("delta.constraints."))
                 .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
     }
 
-    public static boolean changeDataFeedEnabled(MetadataEntry metadataEntry)
+    public static Optional<Boolean> changeDataFeedEnabled(MetadataEntry metadataEntry, ProtocolEntry protocolEntry)
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:543]: Why is it 3-state logic now?
Isn't that CDF is either enabled or not? What does " 🤷‍♂️ " mean here?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 8]
```diff
@@ -491,17 +530,26 @@ private static String getGeneratedColumnExpressions(JsonNode node)
         return generationExpression == null ? null : generationExpression.asText();
     }
 
-    public static Map<String, String> getCheckConstraints(MetadataEntry metadataEntry)
+    public static Map<String, String> getCheckConstraints(MetadataEntry metadataEntry, ProtocolEntry protocolEntry)
     {
+        if (protocolEntry.supportsWriterFeature() && !protocolEntry.writerFeaturesContains("checkConstraints")) {
+            return ImmutableMap.of();
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:536]: Why we check for checkConstraints now and we didn't before?
If the previous logic was flawed, and this is a fix, it would be nice to have it as a separate commit.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 9]
```diff
@@ -457,11 +483,24 @@ private static boolean isIdentityColumn(JsonNode node)
                 .anyMatch(name -> name.startsWith("delta.identity."));
     }
 
-    public static Map<String, String> getColumnInvariants(MetadataEntry metadataEntry)
+    public static Map<String, String> getColumnInvariants(MetadataEntry metadataEntry, ProtocolEntry protocolEntry)
     {
+        if (protocolEntry.supportsWriterFeature()) {
+            if (!protocolEntry.writerFeaturesContains("invariants")) {
+                return ImmutableMap.of();
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:490]: Why we check for checkConstraints now and we didn't before?
If the previous logic was flawed, and this is a fix, it would be nice to have it as a separate commit.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 10]
```diff
@@ -446,8 +469,11 @@ public static Map<String, Boolean> getColumnsNullability(MetadataEntry metadataE
         return getColumnProperties(metadataEntry, node -> node.get("nullable").asBoolean());
     }
 
-    public static Map<String, Boolean> getColumnIdentities(MetadataEntry metadataEntry)
+    public static Map<String, Boolean> getColumnIdentities(MetadataEntry metadataEntry, ProtocolEntry protocolEntry)
     {
+        if (protocolEntry.supportsWriterFeature() && !protocolEntry.writerFeaturesContains("identityColumns")) {
+            return ImmutableMap.of();
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:475]: Why we check for checkConstraints now and we didn't before?
If the previous logic was flawed, and this is a fix, it would be nice to have it as a separate commit.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 11]
```diff
@@ -2180,21 +2206,16 @@ private void checkUnsupportedGeneratedColumns(MetadataEntry metadataEntry)
...
         if (requiredWriterVersion > MAX_WRITER_VERSION) {
             throw new TrinoException(
                     NOT_SUPPORTED,
                     format("Table %s requires Delta Lake writer version %d which is not supported", handle.getSchemaTableName(), requiredWriterVersion));
         }
     }
 
-    private ProtocolEntry getProtocolEntry(ConnectorSession session, DeltaLakeTableHandle handle)
-    {
-        return transactionLogAccess.getProtocolEntry(session, getSnapshot(handle.getSchemaTableName(), handle.getLocation(), session));
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2195]: I like the addition of the protocol entry to the table handle -- especially that `getSnapshot(SchemaTableName)` was getting _current_ snapshot, and not the right snapshot we should be operating on.

Can the addition of the protocol entry to the table handle be a separate commit?
I think it would make the changes more digestible.

**Review by findepi — COMMENTED** [review]
(not a full review yet)

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeFileOperations.java [hunk 12]
```diff
@@ -228,14 +228,14 @@ public void testDeleteWholePartition()
         assertFileSystemAccesses(
                 "DELETE FROM test_delete_part_key WHERE key = 'p1'",
                 ImmutableMultiset.<FileOperation>builder()
-                        .addCopies(new FileOperation(LAST_CHECKPOINT, "_last_checkpoint", INPUT_FILE_NEW_STREAM), 6) // TODO (https://github.com/trinodb/trino/issues/16782) should be checked once per query
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeFileOperations.java:231]: Why did these go down, fewer times reading the Protocol?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 13]
```diff
@@ -2160,16 +2178,24 @@ private boolean allowWrite(ConnectorSession session, DeltaLakeTableHandle tableH
...
         }
-        if (getColumnIdentities(handle.getMetadataEntry()).values().stream().anyMatch(identity -> identity)) {
+        if (getColumnIdentities(handle.getMetadataEntry(), handle.getProtocolEntry()).values().stream().anyMatch(identity -> identity)) {
             throw new TrinoException(NOT_SUPPORTED, "Writing to tables with identity columns is not supported");
         }
-        // TODO: Check writer-features
+        checkUnsupportedWriterFeature(handle.getProtocolEntry());
+    }
+
+    private static void checkUnsupportedWriterFeature(ProtocolEntry protocolEntry)
```
Comment (alexjo2144) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:2193]:
```suggestion
    private static void checkUnsupportedWriterFeatures(ProtocolEntry protocolEntry)
```

**Review by alexjo2144 — COMMENTED** [review]
Just a few naming nit-picks, overall looks pretty good to me

**Comment by ebyhr** [comment]
Rebased on master to resolve conflicts. Comments are not yet addressed.

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeFileOperations.java (hunk 12 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeFileOperations.java:231]: The reduction comes from `checkSupportedWriterVersion` change. The method called `TransactionLogAccess.getProtocolEntry` previously.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java (hunk 8 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:536]: We didn't check because the previous max writer version 6 shouldn't have writer features. The empty writer features in versions <= 6 is ensured in `ProtocolEntry` class. I don't think the previous logic was flawed. 

Same for `invariants` & `identityColumns`.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java (hunk 7 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:543]: Actually, even the previous logic is 3-state because MetadataEntry class had `Optional<Boolean> isChangeDataFeedEnabled()` method. I guess the motivation of the 3-state is hiding `change_data_feed_enabled` table properties in `SHOW CREATE TABLE` when it's undefined. 

cc: @homar

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java [hunk 14]
```diff
@@ -94,6 +96,13 @@ private DeltaLakeSchemaSupport() {}
             .add("columnMapping")
             .add("timestampNtz")
             .build();
+    private static final Set<String> SUPPORTED_WRITER_FEATURES = ImmutableSet.<String>builder()
+            .add("appendOnly")
+            .add("invariants")
+            .add("checkConstraints")
+            .add("changeDataFeed")
+            .add("columnMapping")
```
[the remark this answers is not in the document]
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:104]: Agreed, but let me handle in follow-up.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java (hunk 10 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/DeltaLakeSchemaSupport.java:476]: Separated into "Add ProtocolEntry to DeltaLakeTableHandle" commit.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 4 above)
Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:702]: Yes, it will increase IO in such cases :(

**Comment by ebyhr** [comment]
Addressed comments.

**Comment by ebyhr** [comment]
Rebased on master to resolve conflicts.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 15]
```diff
@@ -563,10 +565,10 @@ public ConnectorTableMetadata getTableMetadata(ConnectorSession session, Connect
         Optional<Long> checkpointInterval = metadataEntry.getCheckpointInterval();
         checkpointInterval.ifPresent(value -> properties.put(CHECKPOINT_INTERVAL_PROPERTY, value));
 
-        Optional<Boolean> changeDataFeedEnabled = metadataEntry.isChangeDataFeedEnabled();
-        changeDataFeedEnabled.ifPresent(value -> properties.put(CHANGE_DATA_FEED_ENABLED_PROPERTY, value));
+        changeDataFeedEnabled(metadataEntry, protocolEntry)
+                .ifPresent(value -> properties.put(CHANGE_DATA_FEED_ENABLED_PROPERTY, value));
 
-        ColumnMappingMode columnMappingMode = DeltaLakeSchemaSupport.getColumnMappingMode(metadataEntry);
+        ColumnMappingMode columnMappingMode = DeltaLakeSchemaSupport.getColumnMappingMode(metadataEntry, protocolEntry);
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:571]: pre-existing: static import - it is already used several other times within this class.

File: plugin/trino-delta-lake/src/test/resources/databricks/invariants_writer_feature/README.md [hunk 16]
```diff
@@ -0,0 +1,29 @@
+Data generated using Databricks 12.2:
+Using PySpark because Spark SQL doesn't support creating a table with column invariants.
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/resources/databricks/invariants_writer_feature/README.md:2]: please place the resource files under `databricks122` directory

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java [hunk 17]
```diff
@@ -199,6 +204,46 @@ public void testCheckConstraintMergeCompatibility()
         }
     }
 
+    @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_OSS, DELTA_LAKE_EXCLUDE_91, DELTA_LAKE_EXCLUDE_104, DELTA_LAKE_EXCLUDE_113, PROFILE_SPECIFIC_TESTS})
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java:207]: This PT could be a delta lake oss only PT.
Do you see any eventual functionality collision potential with DBX?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java [hunk 18]
```diff
@@ -79,6 +83,52 @@ public void testColumnMappingModeNone()
         }
     }
 
+    @Test(
+            groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_OSS, DELTA_LAKE_EXCLUDE_91, DELTA_LAKE_EXCLUDE_104, DELTA_LAKE_EXCLUDE_113, PROFILE_SPECIFIC_TESTS},
+            dataProvider = "supportedColumnMappingForDmlDataProvider")
+    @Flaky(issue = DATABRICKS_COMMUNICATION_FAILURE_ISSUE, match = DATABRICKS_COMMUNICATION_FAILURE_MATCH)
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeColumnMappingMode.java:89]: Run only on delta lake oss / integration test?

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDeleteCompatibility.java [hunk 19]
```diff
@@ -97,6 +99,52 @@ public void testDeleteOnAppendOnlyTableFails()
         onTrino().executeQuery("DROP TABLE " + tableName);
     }
 
+    @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_OSS, DELTA_LAKE_EXCLUDE_91, DELTA_LAKE_EXCLUDE_104, DELTA_LAKE_EXCLUDE_113, PROFILE_SPECIFIC_TESTS})
+    @Flaky(issue = DATABRICKS_COMMUNICATION_FAILURE_ISSUE, match = DATABRICKS_COMMUNICATION_FAILURE_MATCH)
+    public void testDeleteOnAppendOnlyWriterFeature()
+    {
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeDeleteCompatibility.java:105]: Delta Lake OSS / move to integration test

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeWriteDatabricksCompatibility.java [hunk 20]
```diff
@@ -395,17 +396,42 @@ private void testVacuumRemoveChangeDataFeedFiles(Consumer<String> vacuumExecutor
...
+                    .hasMessageContaining("Cannot execute vacuum procedure with 8 writer version");
+        }
+        finally {
+            dropDeltaTableWithRetry("default." + tableName);
+        }
+    }
+
+    @Test(groups = {DELTA_LAKE_DATABRICKS, DELTA_LAKE_EXCLUDE_91, DELTA_LAKE_EXCLUDE_104, DELTA_LAKE_EXCLUDE_113, PROFILE_SPECIFIC_TESTS})
+    @Flaky(issue = DATABRICKS_COMMUNICATION_FAILURE_ISSUE, match = DATABRICKS_COMMUNICATION_FAILURE_MATCH)
+    public void testVacuumUnsupportedWriterFeature()
```
Comment (findinpath) [inline testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeWriteDatabricksCompatibility.java:421]: Delta Lake OSS only

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java [hunk 21]
```diff
@@ -174,30 +176,32 @@ public Stream<DeltaLakeTransactionLogEntry> getCheckpointTransactionLogEntries(
...
-                        session,
-                        checkpointSchemaManager,
-                        typeManager,
-                        fileSystem,
-                        stats,
-                        checkpoint)) :
-                Optional.empty();
+        Optional<MetadataAndProtocolEntry> metadataAndProtocol = Optional.empty();
+        if (entryTypes.contains(ADD)) {
+            metadataAndProtocol = Optional.of(getCheckpointMetadataEntry(
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java]: pre-existing: potential area of improvement we call `getCheckpointTransactionLogEntries` 2 times when dealing with `ADD` - once for metadata & protocol and once for the `entryTypes`. With some extra logic on `CheckpointEntryIterator` we could potentially spare the first checkpoint file access

File: plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeMinioAndHmsConnectorSmokeTest.java [hunk 22]
```diff
@@ -192,6 +192,28 @@ public void testDeltaColumnInvariant()
         assertQuery("SELECT * FROM " + tableName, "VALUES (1), (2)");
     }
 
+    /**
+     * @see databricks.invariants_writer_feature
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/test/java/io/trino/plugin/deltalake/TestDeltaLakeMinioAndHmsConnectorSmokeTest.java:196]: databricks122

File: testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java (hunk 17 above)
Comment (ebyhr) [reply testing/trino-product-tests/src/main/java/io/trino/tests/product/deltalake/TestDeltaLakeCheckConstraintCompatibility.java:207]: I knew it could be OSS only, but I believe there's no such consensus yet.

**Comment by ebyhr** [comment]
Addressed comments.

**Review by findinpath — APPROVED** [review]

**Comment by ebyhr** [comment]
@findepi @alexjo2144 Could you take another look when you have time?

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java [hunk 23]
```diff
@@ -501,7 +502,7 @@ public LocatedTableHandle getTableHandle(ConnectorSession session, SchemaTableNa
             LOG.debug("Skip %s because the table contains unsupported reader features: %s", tableName, unsupportedReaderFeatures);
             return null;
         }
-        verifySupportedColumnMapping(getColumnMappingMode(metadataEntry));
+        verifySupportedColumnMapping(getColumnMappingMode(metadataEntry, protocolEntry));
```
Comment (findepi) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:505]: Can this change go as a separate prep commit?

Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:505]: Even if we separate a commit, the preparatory commit will contain almost all changes except for tests.

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSinkProvider.java [hunk 24]
```diff
@@ -193,7 +193,7 @@ private DeltaLakeCdfPageSink createCdfPageSink(
     {
         MetadataEntry metadataEntry = mergeTableHandle.getTableHandle().getMetadataEntry();
         Set<String> partitionKeys = mergeTableHandle.getTableHandle().getMetadataEntry().getOriginalPartitionColumns().stream().collect(toImmutableSet());
-        List<DeltaLakeColumnHandle> tableColumns = extractSchema(metadataEntry, typeManager).stream()
+        List<DeltaLakeColumnHandle> tableColumns = extractSchema(metadataEntry, mergeTableHandle.getTableHandle().getProtocolEntry(), typeManager).stream()
```
Comment (findinpath) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakePageSinkProvider.java:196]: Extract `mergeTableHandle.getTableHandle().getProtocolEntry()` to a variable because it is being used several times within this method.

**Review by findinpath — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java (hunk 4 above)
Comment (findepi) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/DeltaLakeMetadata.java:702]: will this be solved after https://github.com/trinodb/trino/pull/18919 ?

**Comment by ebyhr** [comment]
Rebased on master to resolve conflicts and added another commit to handle deletion vectors feature in protocol entry.

**Review by findepi — APPROVED** [review]

File: plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java [hunk 25]
```diff
@@ -253,23 +259,46 @@ private MetadataEntry getCheckpointMetadataEntry(
...
     {
+        MetadataEntry metadata = null;
+        ProtocolEntry protocol = null;
         for (Location checkpointPath : getCheckpointPartPaths(checkpoint)) {
             TrinoInputFile checkpointFile = fileSystem.newInputFile(checkpointPath);
-            Stream<DeltaLakeTransactionLogEntry> metadataEntries = getCheckpointTransactionLogEntries(
+            Iterator<DeltaLakeTransactionLogEntry> entries = getCheckpointTransactionLogEntries(
                     session,
-                    ImmutableSet.of(METADATA),
+                    ImmutableSet.of(METADATA, PROTOCOL),
```
Comment (jkylling) [inline plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java:268]: This change likely breaks the checkpoint reading optimizations added in https://github.com/trinodb/trino/commit/06e1339755251675a8104d523be3461b55211dd7

We should probably adopt the optimization to support or-ing multiple tuple domains.

Comment (ebyhr) [reply plugin/trino-delta-lake/src/main/java/io/trino/plugin/deltalake/transactionlog/TableSnapshot.java:268]: Please file an issue to https://github.com/trinodb/trino/issues/new
